From 87f17bfc17608a325f82337f9921ffb165dbc572 Mon Sep 17 00:00:00 2001
From: Anssi Hannula <anssi.hannula@iki.fi>
Date: Fri, 6 Jun 2014 22:58:37 +0300
Subject: [PATCH 1010/1020] ASoC: imx-hdmi: Add support for HBR/HD
 bitstreaming

Enable HBR mode when the userspace sets 8 channel output with non-PCM
indicated in PCM status bits (AES0 & 0x02).

HBR mode allows passthrough/bitstreaming of TrueHD/DTS-HD encoded audio.

ALSA needs an appropriate imx-hdmi-soc.conf for most applications to
automatically set the AESx bits.

Signed-off-by: Anssi Hannula <anssi.hannula@iki.fi>
---
 include/video/mxc_hdmi.h     |    3 ++-
 sound/soc/fsl/imx-hdmi-dma.c |    9 +++++++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/include/video/mxc_hdmi.h b/include/video/mxc_hdmi.h
index dbd0f94f..b5c1e6f 100644
--- a/include/video/mxc_hdmi.h
+++ b/include/video/mxc_hdmi.h
@@ -917,7 +917,8 @@ enum {
 /* AHB_DMA_CONF0 field values */
 	HDMI_AHB_DMA_CONF0_SW_FIFO_RST_OFFSET = 7,
 	HDMI_AHB_DMA_CONF0_SW_FIFO_RST_MASK = 0x80,
-	HDMI_AHB_DMA_CONF0_HBR = 0x10,
+	HDMI_AHB_DMA_CONF0_HBR_OFFSET = 4,
+	HDMI_AHB_DMA_CONF0_HBR_MASK = 0x10,
 	HDMI_AHB_DMA_CONF0_EN_HLOCK_OFFSET = 3,
 	HDMI_AHB_DMA_CONF0_EN_HLOCK_MASK = 0x08,
 	HDMI_AHB_DMA_CONF0_INCR_TYPE_OFFSET = 1,
diff --git a/sound/soc/fsl/imx-hdmi-dma.c b/sound/soc/fsl/imx-hdmi-dma.c
index 4155122..c45e27b 100644
--- a/sound/soc/fsl/imx-hdmi-dma.c
+++ b/sound/soc/fsl/imx-hdmi-dma.c
@@ -785,6 +785,15 @@ static void hdmi_dma_trigger_init(struct snd_pcm_substream *substream,
 				struct hdmi_dma_priv *priv)
 {
 	unsigned long status;
+	bool hbr;
+
+	/*
+	 * Set HBR mode (>192kHz IEC-61937 HD audio bitstreaming).
+	 * This is done this late because userspace may alter the AESx
+	 * parameters until the stream is finally prepared.
+	 */
+	hbr = (iec_header.B.linear_pcm != 0 && priv->channels == 8);
+	hdmi_audio_writeb(AHB_DMA_CONF0, HBR, !!hbr);
 
 	priv->frame_idx = 0;
 
-- 
1.7.9.5

